File: lua/test/magic_eight_ball_entity_test.lua
-- MagicEightBall entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("magic-eight-ball_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("MagicEightBallEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:MagicEightBall(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = magic_eight_ball_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "magic_eight_ball." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set MAGICEIGHTBALL_TEST_MAGIC_EIGHT_BALL_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local magic_eight_ball_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.magic_eight_ball")))
    local magic_eight_ball_ref01_data = nil
    if #magic_eight_ball_ref01_data_raw > 0 then
      magic_eight_ball_ref01_data = helpers.to_map(magic_eight_ball_ref01_data_raw[1][2])
    end

    -- LOAD
    local magic_eight_ball_ref01_ent = client:MagicEightBall(nil)
    local magic_eight_ball_ref01_match_dt0 = {}
    local magic_eight_ball_ref01_data_dt0_loaded, err = magic_eight_ball_ref01_ent:load(magic_eight_ball_ref01_match_dt0, nil)
    assert.is_nil(err)
    assert.is_not_nil(magic_eight_ball_ref01_data_dt0_loaded)

  end)
end)

function magic_eight_ball_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/magic_eight_ball/MagicEightBallTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read magic_eight_ball test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "magic_eight_ball01", "magic_eight_ball02", "magic_eight_ball03", "json01", "json02", "json03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("MAGICEIGHTBALL_TEST_MAGIC_EIGHT_BALL_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["MAGICEIGHTBALL_TEST_MAGIC_EIGHT_BALL_ENTID"] = idmap,
    ["MAGICEIGHTBALL_TEST_LIVE"] = "FALSE",
    ["MAGICEIGHTBALL_TEST_EXPLAIN"] = "FALSE",
    ["MAGICEIGHTBALL_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["MAGICEIGHTBALL_TEST_MAGIC_EIGHT_BALL_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["MAGICEIGHTBALL_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["MAGICEIGHTBALL_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["MAGICEIGHTBALL_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["MAGICEIGHTBALL_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: ts/test/entity/magic_eight_ball/MagicEightBallEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'
import { createLiveTransport } from '../../live-runner'
import { runLiveEntity } from '../../live-entity'


import { MagicEightBallSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('MagicEightBallEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when MAGIC_EIGHT_BALL_TEST_LIVE=TRUE.
  afterEach(liveDelay('MAGIC_EIGHT_BALL_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = MagicEightBallSDK.test()
    const ent = testsdk.MagicEightBall()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.MAGIC_EIGHT_BALL_TEST_LIVE
    for (const op of ['load']) {
      if (!live && maybeSkipControl(t, 'entityOp', 'magic_eight_ball.' + op, live)) return
    }

    
    const setup = basicSetup()
    if (setup.live) {
      return runLiveEntity(setup, {"active":true,"alias":{"field":{}},"fields":[{"active":true,"name":"answer","req":false,"short":"The Magic Eight Ball response","type":"`$STRING`","index$":0},{"active":true,"name":"question","req":false,"short":"The question that was asked","type":"`$STRING`","index$":1},{"active":true,"name":"type","req":false,"short":"The category of the answer (affirmative, non-committal, or negative)","type":"`$STRING`","index$":2}],"name":"magic_eight_ball","op":{"load":{"input":"data","name":"load","points":[{"active":true,"args":{"params":[{"active":true,"example":"Will I be rich?","kind":"param","name":"question","orig":"question","reqd":true,"type":"`$STRING`","index$":0}]},"contract":{"id":"GET /magic/JSON/{question}","json":"{\"operationId\":\"getMagicResponse\",\"parameters\":[{\"description\":\"The question to ask the Magic Eight Ball\",\"example\":\"Will I be rich?\",\"in\":\"path\",\"name\":\"question\",\"required\":true,\"schema\":{\"type\":\"string\"}}],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"example\":{\"magic\":{\"answer\":\"It is certain\",\"question\":\"Will I be rich?\",\"type\":\"affirmative\"}},\"schema\":{\"properties\":{\"magic\":{\"properties\":{\"answer\":{\"description\":\"The Magic Eight Ball response\",\"type\":\"string\"},\"question\":{\"description\":\"The question that was asked\",\"type\":\"string\"},\"type\":{\"description\":\"The category of the answer (affirmative, non-committal, or negative)\",\"enum\":[\"affirmative\",\"non-committal\",\"negative\"],\"type\":\"string\"}},\"type\":\"object\"}},\"type\":\"object\"}}},\"description\":\"Successful response with Magic Eight Ball answer\"},\"400\":{\"description\":\"Bad request - invalid question format\"},\"500\":{\"description\":\"Internal server error\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/magic/JSON/{question}","segments":[{"lit":"magic"},{"lit":"JSON"},{"var":"question"}],"select":{"exist":["question"]},"transform":{"req":"`reqdata`","res":"`body.magic`"},"index$":0}],"key$":"load"}},"relations":{"ancestors":[["json"]]},"key$":"magic_eight_ball","name__orig":"magic_eight_ball","Name":"MagicEightBall","name_":"magic_eight_ball","name-":"magic-eight-ball","NAME":"MAGIC_EIGHT_BALL","index$":0}, {"active":true,"entity":"magic_eight_ball","key$":"BasicMagicEightBallFlow","kind":"basic","name":"BasicMagicEightBallFlow","param":{},"step":[{"active":true,"data":{},"input":{"ref":"magic_eight_ball_ref01","srcdatavar":"magic_eight_ball_ref01_data","suffix":"_dt0"},"match":{"id":"magic_eight_ball01"},"op":"load","spec":[],"valid":[{"apply":"TextFieldMark","def":{"mark":"Mark01-magic_eight_ball_ref01"}}],"index$":0}]}, 'MagicEightBall')
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let magic_eight_ball_ref01_data = Object.values(setup.data.existing.magic_eight_ball)[0] as any

    // LOAD: skipped — no entity id field and load requires path params.
    // Entity-var is declared here so later flow steps still compile.
    const magic_eight_ball_ref01_ent = client.MagicEightBall()


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/magic_eight_ball/MagicEightBallTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = MagicEightBallSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['magic_eight_ball01','magic_eight_ball02','magic_eight_ball03','json01','json02','json03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'MAGIC_EIGHT_BALL_TEST_MAGIC_EIGHT_BALL_ENTID': idmap,
    'MAGIC_EIGHT_BALL_TEST_LIVE': 'FALSE',
    'MAGIC_EIGHT_BALL_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['MAGIC_EIGHT_BALL_TEST_MAGIC_EIGHT_BALL_ENTID']

  const live = 'TRUE' === env.MAGIC_EIGHT_BALL_TEST_LIVE

  const transport = createLiveTransport()
  if (live) {
    const rawIds = process.env['MAGIC_EIGHT_BALL_TEST_MAGIC_EIGHT_BALL_ENTID']
    idmap = rawIds && rawIds.trim() ? JSON.parse(rawIds) : {}
    if (!idmap || Array.isArray(idmap) || typeof idmap !== 'object') {
      throw new Error('Live ENTID must be a JSON object')
    }
    client = new MagicEightBallSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {},
      { system: { fetch: transport.fetch } }
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.MAGIC_EIGHT_BALL_TEST_EXPLAIN,
    live,
    transport,
    now: Date.now(),
  }

  return setup
}
  
